fdbclient/SpecialKeySpace.actor.h

```diff
@@ -38,13 +38,55 @@ class SpecialKeyRangeBaseImpl {
 	// Each derived class only needs to implement this simple version of getRange
 	virtual Future<Standalone<RangeResultRef>> getRange(ReadYourWritesTransaction* ryw, KeyRangeRef kr) const = 0;
 
-	explicit SpecialKeyRangeBaseImpl(KeyRangeRef kr) : range(kr) {}
+	explicit SpecialKeyRangeBaseImpl(KeyRangeRef kr, bool async = false) : range(kr), async(async) {}
 	KeyRangeRef getKeyRange() const { return range; }
+	bool isAsync() const { return async; }
 
 	virtual ~SpecialKeyRangeBaseImpl() {}
 
 protected:
 	KeyRange range; // underlying key range for this function
+	bool async; // true if the range read emits a rpc call, thus we cache the results to keep consistency in the same
+	            // range read
+};
+
+class SpecialKeyRangeAsyncImpl : public SpecialKeyRangeBaseImpl {
+public:
+	explicit SpecialKeyRangeAsyncImpl(KeyRangeRef kr) : SpecialKeyRangeBaseImpl(kr, true) {}
+
+	Future<Standalone<RangeResultRef>> getRange(ReadYourWritesTransaction* ryw, KeyRangeRef kr) const = 0;
+
+	// calling with a cache object to have consistent results if we need to call rpc
+	Future<Standalone<RangeResultRef>> getRange(ReadYourWritesTransaction* ryw, KeyRangeRef kr,
+	                                            Optional<Standalone<RangeResultRef>>* cache) const {
+		return getRangeAsyncActor(this, ryw, kr, cache);
+	}
+
+	ACTOR static Future<Standalone<RangeResultRef>> getRangeAsyncActor(const SpecialKeyRangeBaseImpl* skrAyncImpl,
+	                                                                   ReadYourWritesTransaction* ryw, KeyRangeRef kr,
+	                                                                   Optional<Standalone<RangeResultRef>>* cache) {
+		ASSERT(skrAyncImpl->getKeyRange().contains(kr));
+		if (cache == nullptr) {
```

Comment:
As far as I can tell this is always non null

Reply:
Nice catch! I forgot to change it. Thanks a lot.